ens/utils.py

```diff
@@ -23,38 +39,46 @@
 default = object()
 
 
-def Web3():
-    from web3 import Web3
-    return Web3
+if TYPE_CHECKING:
+    from web3 import Web3 as Web3Type  # noqa: F401
+    from web3.providers import (  # noqa: F401
+        BaseProvider,
+    )
+
+
+def Web3() -> Type['Web3Type']:
+    from web3 import Web3 as Web3Main
+    return Web3Main
 
 
-def dict_copy(func):
+def dict_copy(func: Callable[..., Any]) -> Callable[..., Any]:
     "copy dict keyword args, to avoid modifying caller's copy"
     @functools.wraps(func)
-    def wrapper(*args, **kwargs):
+    def wrapper(*args: Any, **kwargs: Any) -> Callable[..., Any]:
         copied_kwargs = copy.deepcopy(kwargs)
         return func(*args, **copied_kwargs)
     return wrapper
 
 
-def ensure_hex(data):
+def ensure_hex(data: HexBytes) -> HexBytes:
     if not isinstance(data, str):
         return Web3().toHex(data)
     return data
 
 
-def init_web3(providers=default):
-    from web3 import Web3
+# this was providers -> provider?
+def init_web3(provider: 'BaseProvider'=cast('BaseProvider', default)) -> 'Web3Type':
```

Comment:
I changed the arg `providers` -> a single `provider`. If I understand correctly, initializing a `web3` instance only accepts a single `provider` not some kind of sequence or iterable of providers.

Reply:
cc @kclowes

Reply:
Yep, you're correct! We no longer accept multiple providers.